File: src/test/java/com/giwankim/leetcode/MostCommonWordTest.java
package com.giwankim.leetcode;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

class MostCommonWordTest {
  @ParameterizedTest
  @MethodSource
  void mostCommonWord(String paragraph, String[] banned, String expected) {
    String actual = new MostCommonWord().mostCommonWord(paragraph, banned);
    assertThat(actual).isEqualTo(expected);
  }

  private static Stream<Arguments> mostCommonWord() {
    return Stream.of(
        Arguments.of("a.", new String[] {}, "a"),
        Arguments.of(
            "Bob hit a ball, the hit BALL flew far after it was hit.",
            new String[] {"hit"},
            "ball"));
  }
}

File: leetcode-kt/src/test/kotlin/com/giwankim/leetcode/MostCommonWordTest.kt
package com.giwankim.leetcode

import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withTests
import io.kotest.matchers.shouldBe

class MostCommonWordTest :
    FunSpec({
        val sut = MostCommonWord()

        context("most common word") {
            withTests(
                nameFn = { (paragraph, banned, expected) ->
                    "paragraph=${paragraph.take(20)}..., banned=${banned.contentToString()}, expected=$expected"
                },
                MostCommonWordCase("a.", emptyArray(), "a"),
                MostCommonWordCase("Bob hit a ball, the hit BALL flew far after it was hit.", arrayOf("hit"), "ball"),
            ) { (paragraph, banned, expected) ->
                sut.mostCommonWord(paragraph, banned) shouldBe expected
            }
        }
    })

private data class MostCommonWordCase(
    val paragraph: String,
    val banned: Array<String>,
    val expected: String,
)
